// vite-plus@0.1.13/dist/global/package-D0VpX-8d.js
import path from "node:path";
import fs from "node:fs";
import { createRequire } from "node:module";
//#region src/utils/constants.ts
const VITE_PLUS_NAME = "vite-plus";
const VITE_PLUS_VERSION = process.env.VITE_PLUS_VERSION || "latest";
const VITE_PLUS_OVERRIDE_PACKAGES = process.env.VITE_PLUS_OVERRIDE_PACKAGES
  ? JSON.parse(process.env.VITE_PLUS_OVERRIDE_PACKAGES)
  : {
      vite: "npm:@voidzero-dev/vite-plus-core@latest",
      vitest: "npm:@voidzero-dev/vite-plus-test@latest",
    };
createRequire(import.meta.url);
const BASEURL_TSCONFIG_WARNING =
  "Skipped typeAware/typeCheck: tsconfig.json contains baseUrl which is not yet supported by the oxlint type checker.\n  Run `npx @andrewbranch/ts5to6 --fixBaseUrl .` to remove baseUrl from your tsconfig.";
(process.versions.node, process.release.name);
//#endregion
//#region ../../node_modules/.pnpm/detect-indent@7.0.2/node_modules/detect-indent/index.js
const INDENT_REGEX = /^(?:( )+|\t+)/;
const INDENT_TYPE_SPACE = "space";
const INDENT_TYPE_TAB = "tab";
function shouldIgnoreSingleSpace(ignoreSingleSpaces, indentType, value) {
  return ignoreSingleSpaces && indentType === INDENT_TYPE_SPACE && value === 1;
}
/**
Make a Map that counts how many indents/unindents have occurred for a given size and how many lines follow a given indentation.

The key is a concatenation of the indentation type (s = space and t = tab) and the size of the indents/unindents.

```
indents = {
t3: [1, 0],
t4: [1, 5],
s5: [1, 0],
s12: [1, 0],
}
```
*/
function makeIndentsMap(string, ignoreSingleSpaces) {
  const indents = /* @__PURE__ */ new Map();
  let previousSize = 0;
  let previousIndentType;
  let key;
  for (const line of string.split(/\n/g)) {
    if (!line) continue;
    const matches = line.match(INDENT_REGEX);
    if (matches === null) {
      previousSize = 0;
      previousIndentType = "";
    } else {
      const indent = matches[0].length;
      const indentType = matches[1] ? INDENT_TYPE_SPACE : INDENT_TYPE_TAB;
      if (shouldIgnoreSingleSpace(ignoreSingleSpaces, indentType, indent)) continue;
      if (indentType !== previousIndentType) previousSize = 0;
      previousIndentType = indentType;
      let use = 1;
      let weight = 0;
      const indentDifference = indent - previousSize;
      previousSize = indent;
      if (indentDifference === 0) {
        use = 0;
        weight = 1;
      } else {
        const absoluteIndentDifference = Math.abs(indentDifference);
        if (shouldIgnoreSingleSpace(ignoreSingleSpaces, indentType, absoluteIndentDifference))
          continue;
        key = encodeIndentsKey(indentType, absoluteIndentDifference);
      }
      const entry = indents.get(key);
      indents.set(key, entry === void 0 ? [1, 0] : [entry[0] + use, entry[1] + weight]);
    }
  }
  return indents;
}
function encodeIndentsKey(indentType, indentAmount) {
  return (indentType === INDENT_TYPE_SPACE ? "s" : "t") + String(indentAmount);
}
function decodeIndentsKey(indentsKey) {
  return {
    type: indentsKey[0] === "s" ? INDENT_TYPE_SPACE : INDENT_TYPE_TAB,
    amount: Number(indentsKey.slice(1)),
  };
}
function getMostUsedKey(indents) {
  let result;
  let maxUsed = 0;
  let maxWeight = 0;
  for (const [key, [usedCount, weight]] of indents)
    if (usedCount > maxUsed || (usedCount === maxUsed && weight > maxWeight)) {
      maxUsed = usedCount;
      maxWeight = weight;
      result = key;
    }
  return result;
}
function makeIndentString(type, amount) {
  return (type === INDENT_TYPE_SPACE ? " " : "	").repeat(amount);
}
function detectIndent(string) {
  if (typeof string !== "string") throw new TypeError("Expected a string");
  let indents = makeIndentsMap(string, true);
  if (indents.size === 0) indents = makeIndentsMap(string, false);
  const keyOfMostUsedIndent = getMostUsedKey(indents);
  let type;
  let amount = 0;
  let indent = "";
  if (keyOfMostUsedIndent !== void 0) {
    ({ type, amount } = decodeIndentsKey(keyOfMostUsedIndent));
    indent = makeIndentString(type, amount);
  }
  return {
    amount,
    type,
    indent,
  };
}
//#endregion
//#region ../../node_modules/.pnpm/detect-newline@4.0.1/node_modules/detect-newline/index.js
function detectNewline(string) {
  if (typeof string !== "string") throw new TypeError("Expected a string");
  const newlines = string.match(/(?:\r?\n)/g) || [];
  if (newlines.length === 0) return;
  const crlf = newlines.filter((newline) => newline === "\r\n").length;
  return crlf > newlines.length - crlf ? "\r\n" : "\n";
}
//#endregion
//#region ../../node_modules/.pnpm/jsonc-parser@3.3.1/node_modules/jsonc-parser/lib/esm/impl/scanner.js
/**
 * Creates a JSON scanner on the given text.
 * If ignoreTrivia is set, whitespaces or comments are ignored.
 */
function createScanner$1(text, ignoreTrivia = false) {
  const len = text.length;
  let pos = 0,
    value = "",
    tokenOffset = 0,
    token = 16,
    lineNumber = 0,
    lineStartOffset = 0,
    tokenLineStartOffset = 0,
    prevTokenLineStartOffset = 0,
    scanError = 0;
  function scanHexDigits(count, exact) {
    let digits = 0;
    let value = 0;
    while (digits < count || !exact) {
      let ch = text.charCodeAt(pos);
      if (ch >= 48 && ch <= 57) value = value * 16 + ch - 48;
      else if (ch >= 65 && ch <= 70) value = value * 16 + ch - 65 + 10;
      else if (ch >= 97 && ch <= 102) value = value * 16 + ch - 97 + 10;
      else break;
      pos++;
      digits++;
    }
    if (digits < count) value = -1;
    return value;
  }
  function setPosition(newPosition) {
    pos = newPosition;
    value = "";
    tokenOffset = 0;
    token = 16;
    scanError = 0;
  }
  function scanNumber() {
    let start = pos;
    if (text.charCodeAt(pos) === 48) pos++;
    else {
      pos++;
      while (pos < text.length && isDigit(text.charCodeAt(pos))) pos++;
    }
    if (pos < text.length && text.charCodeAt(pos) === 46) {
      pos++;
      if (pos < text.length && isDigit(text.charCodeAt(pos))) {
        pos++;
        while (pos < text.length && isDigit(text.charCodeAt(pos))) pos++;
      } else {
        scanError = 3;
        return text.substring(start, pos);
      }
    }
    let end = pos;
    if (pos < text.length && (text.charCodeAt(pos) === 69 || text.charCodeAt(pos) === 101)) {
      pos++;
      if ((pos < text.length && text.charCodeAt(pos) === 43) || text.charCodeAt(pos) === 45) pos++;
      if (pos < text.length && isDigit(text.charCodeAt(pos))) {
        pos++;
        while (pos < text.length && isDigit(text.charCodeAt(pos))) pos++;
        end = pos;
      } else scanError = 3;
    }
    return text.substring(start, end);
  }
  function scanString() {
    let result = "",
      start = pos;
    while (true) {
      if (pos >= len) {
        result += text.substring(start, pos);
        scanError = 2;
        break;
      }
      const ch = text.charCodeAt(pos);
      if (ch === 34) {
        result += text.substring(start, pos);
        pos++;
        break;
      }
      if (ch === 92) {
        result += text.substring(start, pos);
        pos++;
        if (pos >= len) {
          scanError = 2;
          break;
        }
        switch (text.charCodeAt(pos++)) {
          case 34:
            result += '"';
            break;
          case 92:
            result += "\\";
            break;
          case 47:
            result += "/";
            break;
          case 98:
            result += "\b";
            break;
          case 102:
            result += "\f";
            break;
          case 110:
            result += "\n";
            break;
          case 114:
            result += "\r";
            break;
          case 116:
            result += "	";
            break;
          case 117:
            const ch3 = scanHexDigits(4, true);
            if (ch3 >= 0) result += String.fromCharCode(ch3);
            else scanError = 4;
            break;
          default:
            scanError = 5;
        }
        start = pos;
        continue;
      }
      if (ch >= 0 && ch <= 31)
        if (isLineBreak(ch)) {
          result += text.substring(start, pos);
          scanError = 2;
          break;
        } else scanError = 6;
      pos++;
    }
    return result;
  }
  function scanNext() {
    value = "";
    scanError = 0;
    tokenOffset = pos;
    lineStartOffset = lineNumber;
    prevTokenLineStartOffset = tokenLineStartOffset;
    if (pos >= len) {
      tokenOffset = len;
      return (token = 17);
    }
    let code = text.charCodeAt(pos);
    if (isWhiteSpace(code)) {
      do {
        pos++;
        value += String.fromCharCode(code);
        code = text.charCodeAt(pos);
      } while (isWhiteSpace(code));
      return (token = 15);
    }
    if (isLineBreak(code)) {
      pos++;
      value += String.fromCharCode(code);
      if (code === 13 && text.charCodeAt(pos) === 10) {
        pos++;
        value += "\n";
      }
      lineNumber++;
      tokenLineStartOffset = pos;
      return (token = 14);
    }
    switch (code) {
      case 123:
        pos++;
        return (token = 1);
      case 125:
        pos++;
        return (token = 2);
      case 91:
        pos++;
        return (token = 3);
      case 93:
        pos++;
        return (token = 4);
      case 58:
        pos++;
        return (token = 6);
      case 44:
        pos++;
        return (token = 5);
      case 34:
        pos++;
        value = scanString();
        return (token = 10);
      case 47:
        const start = pos - 1;
        if (text.charCodeAt(pos + 1) === 47) {
          pos += 2;
          while (pos < len) {
            if (isLineBreak(text.charCodeAt(pos))) break;
            pos++;
          }
          value = text.substring(start, pos);
          return (token = 12);
        }
        if (text.charCodeAt(pos + 1) === 42) {
          pos += 2;
          const safeLength = len - 1;
          let commentClosed = false;
          while (pos < safeLength) {
            const ch = text.charCodeAt(pos);
            if (ch === 42 && text.charCodeAt(pos + 1) === 47) {
              pos += 2;
              commentClosed = true;
              break;
            }
            pos++;
            if (isLineBreak(ch)) {
              if (ch === 13 && text.charCodeAt(pos) === 10) pos++;
              lineNumber++;
              tokenLineStartOffset = pos;
            }
          }
          if (!commentClosed) {
            pos++;
            scanError = 1;
          }
          value = text.substring(start, pos);
          return (token = 13);
        }
        value += String.fromCharCode(code);
        pos++;
        return (token = 16);
      case 45:
        value += String.fromCharCode(code);
        pos++;
        if (pos === len || !isDigit(text.charCodeAt(pos))) return (token = 16);
      case 48:
      case 49:
      case 50:
      case 51:
      case 52:
      case 53:
      case 54:
      case 55:
      case 56:
      case 57:
        value += scanNumber();
        return (token = 11);
      default:
        while (pos < len && isUnknownContentCharacter(code)) {
          pos++;
          code = text.charCodeAt(pos);
        }
        if (tokenOffset !== pos) {
          value = text.substring(tokenOffset, pos);
          switch (value) {
            case "true":
              return (token = 8);
            case "false":
              return (token = 9);
            case "null":
              return (token = 7);
          }
          return (token = 16);
        }
        value += String.fromCharCode(code);
        pos++;
        return (token = 16);
    }
  }
  function isUnknownContentCharacter(code) {
    if (isWhiteSpace(code) || isLineBreak(code)) return false;
    switch (code) {
      case 125:
      case 93:
      case 123:
      case 91:
      case 34:
      case 58:
      case 44:
      case 47:
        return false;
    }
    return true;
  }
  function scanNextNonTrivia() {
    let result;
    do result = scanNext();
    while (result >= 12 && result <= 15);
    return result;
  }
  return {
    setPosition,
    getPosition: () => pos,
    scan: ignoreTrivia ? scanNextNonTrivia : scanNext,
    getToken: () => token,
    getTokenValue: () => value,
    getTokenOffset: () => tokenOffset,
    getTokenLength: () => pos - tokenOffset,
    getTokenStartLine: () => lineStartOffset,
    getTokenStartCharacter: () => tokenOffset - prevTokenLineStartOffset,
    getTokenError: () => scanError,
  };
}
function isWhiteSpace(ch) {
  return ch === 32 || ch === 9;
}
function isLineBreak(ch) {
  return ch === 10 || ch === 13;
}
function isDigit(ch) {
  return ch >= 48 && ch <= 57;
}
var CharacterCodes;
(function (CharacterCodes) {
  CharacterCodes[(CharacterCodes["lineFeed"] = 10)] = "lineFeed";
  CharacterCodes[(CharacterCodes["carriageReturn"] = 13)] = "carriageReturn";
  CharacterCodes[(CharacterCodes["space"] = 32)] = "space";
  CharacterCodes[(CharacterCodes["_0"] = 48)] = "_0";
  CharacterCodes[(CharacterCodes["_1"] = 49)] = "_1";
  CharacterCodes[(CharacterCodes["_2"] = 50)] = "_2";
  CharacterCodes[(CharacterCodes["_3"] = 51)] = "_3";
  CharacterCodes[(CharacterCodes["_4"] = 52)] = "_4";
  CharacterCodes[(CharacterCodes["_5"] = 53)] = "_5";
  CharacterCodes[(CharacterCodes["_6"] = 54)] = "_6";
  CharacterCodes[(CharacterCodes["_7"] = 55)] = "_7";
  CharacterCodes[(CharacterCodes["_8"] = 56)] = "_8";
  CharacterCodes[(CharacterCodes["_9"] = 57)] = "_9";
  CharacterCodes[(CharacterCodes["a"] = 97)] = "a";
  CharacterCodes[(CharacterCodes["b"] = 98)] = "b";
  CharacterCodes[(CharacterCodes["c"] = 99)] = "c";
  CharacterCodes[(CharacterCodes["d"] = 100)] = "d";
  CharacterCodes[(CharacterCodes["e"] = 101)] = "e";
  CharacterCodes[(CharacterCodes["f"] = 102)] = "f";
  CharacterCodes[(CharacterCodes["g"] = 103)] = "g";
  CharacterCodes[(CharacterCodes["h"] = 104)] = "h";
  CharacterCodes[(CharacterCodes["i"] = 105)] = "i";
  CharacterCodes[(CharacterCodes["j"] = 106)] = "j";
  CharacterCodes[(CharacterCodes["k"] = 107)] = "k";
  CharacterCodes[(CharacterCodes["l"] = 108)] = "l";
  CharacterCodes[(CharacterCodes["m"] = 109)] = "m";
  CharacterCodes[(CharacterCodes["n"] = 110)] = "n";
  CharacterCodes[(CharacterCodes["o"] = 111)] = "o";
  CharacterCodes[(CharacterCodes["p"] = 112)] = "p";
  CharacterCodes[(CharacterCodes["q"] = 113)] = "q";
  CharacterCodes[(CharacterCodes["r"] = 114)] = "r";
  CharacterCodes[(CharacterCodes["s"] = 115)] = "s";
  CharacterCodes[(CharacterCodes["t"] = 116)] = "t";
  CharacterCodes[(CharacterCodes["u"] = 117)] = "u";
  CharacterCodes[(CharacterCodes["v"] = 118)] = "v";
  CharacterCodes[(CharacterCodes["w"] = 119)] = "w";
  CharacterCodes[(CharacterCodes["x"] = 120)] = "x";
  CharacterCodes[(CharacterCodes["y"] = 121)] = "y";
  CharacterCodes[(CharacterCodes["z"] = 122)] = "z";
  CharacterCodes[(CharacterCodes["A"] = 65)] = "A";
  CharacterCodes[(CharacterCodes["B"] = 66)] = "B";
  CharacterCodes[(CharacterCodes["C"] = 67)] = "C";
  CharacterCodes[(CharacterCodes["D"] = 68)] = "D";
  CharacterCodes[(CharacterCodes["E"] = 69)] = "E";
  CharacterCodes[(CharacterCodes["F"] = 70)] = "F";
  CharacterCodes[(CharacterCodes["G"] = 71)] = "G";
  CharacterCodes[(CharacterCodes["H"] = 72)] = "H";
  CharacterCodes[(CharacterCodes["I"] = 73)] = "I";
  CharacterCodes[(CharacterCodes["J"] = 74)] = "J";
  CharacterCodes[(CharacterCodes["K"] = 75)] = "K";
  CharacterCodes[(CharacterCodes["L"] = 76)] = "L";
  CharacterCodes[(CharacterCodes["M"] = 77)] = "M";
  CharacterCodes[(CharacterCodes["N"] = 78)] = "N";
  CharacterCodes[(CharacterCodes["O"] = 79)] = "O";
  CharacterCodes[(CharacterCodes["P"] = 80)] = "P";
  CharacterCodes[(CharacterCodes["Q"] = 81)] = "Q";
  CharacterCodes[(CharacterCodes["R"] = 82)] = "R";
  CharacterCodes[(CharacterCodes["S"] = 83)] = "S";
  CharacterCodes[(CharacterCodes["T"] = 84)] = "T";
  CharacterCodes[(CharacterCodes["U"] = 85)] = "U";
  CharacterCodes[(CharacterCodes["V"] = 86)] = "V";
  CharacterCodes[(CharacterCodes["W"] = 87)] = "W";
  CharacterCodes[(CharacterCodes["X"] = 88)] = "X";
  CharacterCodes[(CharacterCodes["Y"] = 89)] = "Y";
  CharacterCodes[(CharacterCodes["Z"] = 90)] = "Z";
  CharacterCodes[(CharacterCodes["asterisk"] = 42)] = "asterisk";
  CharacterCodes[(CharacterCodes["backslash"] = 92)] = "backslash";
  CharacterCodes[(CharacterCodes["closeBrace"] = 125)] = "closeBrace";
  CharacterCodes[(CharacterCodes["closeBracket"] = 93)] = "closeBracket";
  CharacterCodes[(CharacterCodes["colon"] = 58)] = "colon";
  CharacterCodes[(CharacterCodes["comma"] = 44)] = "comma";
  CharacterCodes[(CharacterCodes["dot"] = 46)] = "dot";
  CharacterCodes[(CharacterCodes["doubleQuote"] = 34)] = "doubleQuote";
  CharacterCodes[(CharacterCodes["minus"] = 45)] = "minus";
  CharacterCodes[(CharacterCodes["openBrace"] = 123)] = "openBrace";
  CharacterCodes[(CharacterCodes["openBracket"] = 91)] = "openBracket";
  CharacterCodes[(CharacterCodes["plus"] = 43)] = "plus";
  CharacterCodes[(CharacterCodes["slash"] = 47)] = "slash";
  CharacterCodes[(CharacterCodes["formFeed"] = 12)] = "formFeed";
  CharacterCodes[(CharacterCodes["tab"] = 9)] = "tab";
})(CharacterCodes || (CharacterCodes = {}));
new Array(20).fill(0).map((_, index) => {
  return " ".repeat(index);
});
const maxCachedValues = 200;
(new Array(maxCachedValues).fill(0).map((_, index) => {
  return "\n" + " ".repeat(index);
}),
  new Array(maxCachedValues).fill(0).map((_, index) => {
    return "\r" + " ".repeat(index);
  }),
  new Array(maxCachedValues).fill(0).map((_, index) => {
    return "\r\n" + " ".repeat(index);
  }),
  new Array(maxCachedValues).fill(0).map((_, index) => {
    return "\n" + "	".repeat(index);
  }),
  new Array(maxCachedValues).fill(0).map((_, index) => {
    return "\r" + "	".repeat(index);
  }),
  new Array(maxCachedValues).fill(0).map((_, index) => {
    return "\r\n" + "	".repeat(index);
  }));
//#endregion
//#region ../../node_modules/.pnpm/jsonc-parser@3.3.1/node_modules/jsonc-parser/lib/esm/impl/parser.js
var ParseOptions;
(function (ParseOptions) {
  ParseOptions.DEFAULT = { allowTrailingComma: false };
})(ParseOptions || (ParseOptions = {}));
/**
 * Parses the given text and returns the object the JSON content represents. On invalid input, the parser tries to be as fault tolerant as possible, but still return a result.
 * Therefore always check the errors list to find out if the input was valid.
 */
function parse$1(text, errors = [], options = ParseOptions.DEFAULT) {
  let currentProperty = null;
  let currentParent = [];
  const previousParents = [];
  function onValue(value) {
    if (Array.isArray(currentParent)) currentParent.push(value);
    else if (currentProperty !== null) currentParent[currentProperty] = value;
  }
  visit$1(
    text,
    {
      onObjectBegin: () => {
        const object = {};
        onValue(object);
        previousParents.push(currentParent);
        currentParent = object;
        currentProperty = null;
      },
      onObjectProperty: (name) => {
        currentProperty = name;
      },
      onObjectEnd: () => {
        currentParent = previousParents.pop();
      },
      onArrayBegin: () => {
        const array = [];
        onValue(array);
        previousParents.push(currentParent);
        currentParent = array;
        currentProperty = null;
      },
      onArrayEnd: () => {
        currentParent = previousParents.pop();
      },
      onLiteralValue: onValue,
      onError: (error, offset, length) => {
        errors.push({
          error,
          offset,
          length,
        });
      },
    },
    options,
  );
  return currentParent[0];
}
/**
 * Gets the JSON path of the given JSON DOM node
 */
function getNodePath$1(node) {
  if (!node.parent || !node.parent.children) return [];
  const path = getNodePath$1(node.parent);
  if (node.parent.type === "property") {
    const key = node.parent.children[0].value;
    path.push(key);
  } else if (node.parent.type === "array") {
    const index = node.parent.children.indexOf(node);
    if (index !== -1) path.push(index);
  }
  return path;
}
/**
 * Evaluates the JavaScript object of the given JSON DOM node
 */
function getNodeValue$1(node) {
  switch (node.type) {
    case "array":
      return node.children.map(getNodeValue$1);
    case "object":
      const obj = Object.create(null);
      for (let prop of node.children) {
        const valueNode = prop.children[1];
        if (valueNode) obj[prop.children[0].value] = getNodeValue$1(valueNode);
      }
      return obj;
    case "null":
    case "string":
    case "number":
    case "boolean":
      return node.value;
    default:
      return;
  }
}
function contains(node, offset, includeRightBound = false) {
  return (
    (offset >= node.offset && offset < node.offset + node.length) ||
    (includeRightBound && offset === node.offset + node.length)
  );
}
/**
 * Finds the most inner node at the given offset. If includeRightBound is set, also finds nodes that end at the given offset.
 */
function findNodeAtOffset$1(node, offset, includeRightBound = false) {
  if (contains(node, offset, includeRightBound)) {
    const children = node.children;
    if (Array.isArray(children))
      for (let i = 0; i < children.length && children[i].offset <= offset; i++) {
        const item = findNodeAtOffset$1(children[i], offset, includeRightBound);
        if (item) return item;
      }
    return node;
  }
}
/**
 * Parses the given text and invokes the visitor functions for each object, array and literal reached.
 */
function visit$1(text, visitor, options = ParseOptions.DEFAULT) {
  const _scanner = createScanner$1(text, false);
  const _jsonPath = [];
  let suppressedCallbacks = 0;
  function toNoArgVisit(visitFunction) {
    return visitFunction
      ? () =>
          suppressedCallbacks === 0 &&
          visitFunction(
            _scanner.getTokenOffset(),
            _scanner.getTokenLength(),
            _scanner.getTokenStartLine(),
            _scanner.getTokenStartCharacter(),
          )
      : () => true;
  }
  function toOneArgVisit(visitFunction) {
    return visitFunction
      ? (arg) =>
          suppressedCallbacks === 0 &&
          visitFunction(
            arg,
            _scanner.getTokenOffset(),
            _scanner.getTokenLength(),
            _scanner.getTokenStartLine(),
            _scanner.getTokenStartCharacter(),
          )
      : () => true;
  }
  function toOneArgVisitWithPath(visitFunction) {
    return visitFunction
      ? (arg) =>
          suppressedCallbacks === 0 &&
          visitFunction(
            arg,
            _scanner.getTokenOffset(),
            _scanner.getTokenLength(),
            _scanner.getTokenStartLine(),
            _scanner.getTokenStartCharacter(),
            () => _jsonPath.slice(),
          )
      : () => true;
  }
  function toBeginVisit(visitFunction) {
    return visitFunction
      ? () => {
          if (suppressedCallbacks > 0) suppressedCallbacks++;
          else if (
            visitFunction(
              _scanner.getTokenOffset(),
              _scanner.getTokenLength(),
              _scanner.getTokenStartLine(),
              _scanner.getTokenStartCharacter(),
              () => _jsonPath.slice(),
            ) === false
          )
            suppressedCallbacks = 1;
        }
      : () => true;
  }
  function toEndVisit(visitFunction) {
    return visitFunction
      ? () => {
          if (suppressedCallbacks > 0) suppressedCallbacks--;
          if (suppressedCallbacks === 0)
            visitFunction(
              _scanner.getTokenOffset(),
              _scanner.getTokenLength(),
              _scanner.getTokenStartLine(),
              _scanner.getTokenStartCharacter(),
            );
        }
      : () => true;
  }
  const onObjectBegin = toBeginVisit(visitor.onObjectBegin),
    onObjectProperty = toOneArgVisitWithPath(visitor.onObjectProperty),
    onObjectEnd = toEndVisit(visitor.onObjectEnd),
    onArrayBegin = toBeginVisit(visitor.onArrayBegin),
    onArrayEnd = toEndVisit(visitor.onArrayEnd),
    onLiteralValue = toOneArgVisitWithPath(visitor.onLiteralValue),
    onSeparator = toOneArgVisit(visitor.onSeparator),
    onComment = toNoArgVisit(visitor.onComment),
    onError = toOneArgVisit(visitor.onError);
  const disallowComments = options && options.disallowComments;
  const allowTrailingComma = options && options.allowTrailingComma;
  function scanNext() {
    while (true) {
      const token = _scanner.scan();
      switch (_scanner.getTokenError()) {
        case 4:
          handleError(14);
          break;
        case 5:
          handleError(15);
          break;
        case 3:
          handleError(13);
          break;
        case 1:
          if (!disallowComments) handleError(11);
          break;
        case 2:
          handleError(12);
          break;
        case 6:
          handleError(16);
          break;
      }
      switch (token) {
        case 12:
        case 13:
          if (disallowComments) handleError(10);
          else onComment();
          break;
        case 16:
          handleError(1);
          break;
        case 15:
        case 14:
          break;
        default:
          return token;
      }
    }
  }
  function handleError(error, skipUntilAfter = [], skipUntil = []) {
    onError(error);
    if (skipUntilAfter.length + skipUntil.length > 0) {
      let token = _scanner.getToken();
      while (token !== 17) {
        if (skipUntilAfter.indexOf(token) !== -1) {
          scanNext();
          break;
        } else if (skipUntil.indexOf(token) !== -1) break;
        token = scanNext();
      }
    }
  }
  function parseString(isValue) {
    const value = _scanner.getTokenValue();
    if (isValue) onLiteralValue(value);
    else {
      onObjectProperty(value);
      _jsonPath.push(value);
    }
    scanNext();
    return true;
  }
  function parseLiteral() {
    switch (_scanner.getToken()) {
      case 11:
        const tokenValue = _scanner.getTokenValue();
        let value = Number(tokenValue);
        if (isNaN(value)) {
          handleError(2);
          value = 0;
        }
        onLiteralValue(value);
        break;
      case 7:
        onLiteralValue(null);
        break;
      case 8:
        onLiteralValue(true);
        break;
      case 9:
        onLiteralValue(false);
        break;
      default:
        return false;
    }
    scanNext();
    return true;
  }
  function parseProperty() {
    if (_scanner.getToken() !== 10) {
      handleError(3, [], [2, 5]);
      return false;
    }
    parseString(false);
    if (_scanner.getToken() === 6) {
      onSeparator(":");
      scanNext();
      if (!parseValue()) handleError(4, [], [2, 5]);
    } else handleError(5, [], [2, 5]);
    _jsonPath.pop();
    return true;
  }
  function parseObject() {
    onObjectBegin();
    scanNext();
    let needsComma = false;
    while (_scanner.getToken() !== 2 && _scanner.getToken() !== 17) {
      if (_scanner.getToken() === 5) {
        if (!needsComma) handleError(4, [], []);
        onSeparator(",");
        scanNext();
        if (_scanner.getToken() === 2 && allowTrailingComma) break;
      } else if (needsComma) handleError(6, [], []);
      if (!parseProperty()) handleError(4, [], [2, 5]);
      needsComma = true;
    }
    onObjectEnd();
    if (_scanner.getToken() !== 2) handleError(7, [2], []);
    else scanNext();
    return true;
  }
  function parseArray() {
    onArrayBegin();
    scanNext();
    let isFirstElement = true;
    let needsComma = false;
    while (_scanner.getToken() !== 4 && _scanner.getToken() !== 17) {
      if (_scanner.getToken() === 5) {
        if (!needsComma) handleError(4, [], []);
        onSeparator(",");
        scanNext();
        if (_scanner.getToken() === 4 && allowTrailingComma) break;
      } else if (needsComma) handleError(6, [], []);
      if (isFirstElement) {
        _jsonPath.push(0);
        isFirstElement = false;
      } else _jsonPath[_jsonPath.length - 1]++;
      if (!parseValue()) handleError(4, [], [4, 5]);
      needsComma = true;
    }
    onArrayEnd();
    if (!isFirstElement) _jsonPath.pop();
    if (_scanner.getToken() !== 4) handleError(8, [4], []);
    else scanNext();
    return true;
  }
  function parseValue() {
    switch (_scanner.getToken()) {
      case 3:
        return parseArray();
      case 1:
        return parseObject();
      case 10:
        return parseString(true);
      default:
        return parseLiteral();
    }
  }
  scanNext();
  if (_scanner.getToken() === 17) {
    if (options.allowEmptyContent) return true;
    handleError(4, [], []);
    return false;
  }
  if (!parseValue()) {
    handleError(4, [], []);
    return false;
  }
  if (_scanner.getToken() !== 17) handleError(9, [], []);
  return true;
}
var ScanError;
(function (ScanError) {
  ScanError[(ScanError["None"] = 0)] = "None";
  ScanError[(ScanError["UnexpectedEndOfComment"] = 1)] = "UnexpectedEndOfComment";
  ScanError[(ScanError["UnexpectedEndOfString"] = 2)] = "UnexpectedEndOfString";
  ScanError[(ScanError["UnexpectedEndOfNumber"] = 3)] = "UnexpectedEndOfNumber";
  ScanError[(ScanError["InvalidUnicode"] = 4)] = "InvalidUnicode";
  ScanError[(ScanError["InvalidEscapeCharacter"] = 5)] = "InvalidEscapeCharacter";
  ScanError[(ScanError["InvalidCharacter"] = 6)] = "InvalidCharacter";
})(ScanError || (ScanError = {}));
var SyntaxKind;
(function (SyntaxKind) {
  SyntaxKind[(SyntaxKind["OpenBraceToken"] = 1)] = "OpenBraceToken";
  SyntaxKind[(SyntaxKind["CloseBraceToken"] = 2)] = "CloseBraceToken";
  SyntaxKind[(SyntaxKind["OpenBracketToken"] = 3)] = "OpenBracketToken";
  SyntaxKind[(SyntaxKind["CloseBracketToken"] = 4)] = "CloseBracketToken";
  SyntaxKind[(SyntaxKind["CommaToken"] = 5)] = "CommaToken";
  SyntaxKind[(SyntaxKind["ColonToken"] = 6)] = "ColonToken";
  SyntaxKind[(SyntaxKind["NullKeyword"] = 7)] = "NullKeyword";
  SyntaxKind[(SyntaxKind["TrueKeyword"] = 8)] = "TrueKeyword";
  SyntaxKind[(SyntaxKind["FalseKeyword"] = 9)] = "FalseKeyword";
  SyntaxKind[(SyntaxKind["StringLiteral"] = 10)] = "StringLiteral";
  SyntaxKind[(SyntaxKind["NumericLiteral"] = 11)] = "NumericLiteral";
  SyntaxKind[(SyntaxKind["LineCommentTrivia"] = 12)] = "LineCommentTrivia";
  SyntaxKind[(SyntaxKind["BlockCommentTrivia"] = 13)] = "BlockCommentTrivia";
  SyntaxKind[(SyntaxKind["LineBreakTrivia"] = 14)] = "LineBreakTrivia";
  SyntaxKind[(SyntaxKind["Trivia"] = 15)] = "Trivia";
  SyntaxKind[(SyntaxKind["Unknown"] = 16)] = "Unknown";
  SyntaxKind[(SyntaxKind["EOF"] = 17)] = "EOF";
})(SyntaxKind || (SyntaxKind = {}));
/**
 * Parses the given text and returns the object the JSON content represents. On invalid input, the parser tries to be as fault tolerant as possible, but still return a result.
 * Therefore, always check the errors list to find out if the input was valid.
 */
const parse = parse$1;
var ParseErrorCode;
(function (ParseErrorCode) {
  ParseErrorCode[(ParseErrorCode["InvalidSymbol"] = 1)] = "InvalidSymbol";
  ParseErrorCode[(ParseErrorCode["InvalidNumberFormat"] = 2)] = "InvalidNumberFormat";
  ParseErrorCode[(ParseErrorCode["PropertyNameExpected"] = 3)] = "PropertyNameExpected";
  ParseErrorCode[(ParseErrorCode["ValueExpected"] = 4)] = "ValueExpected";
  ParseErrorCode[(ParseErrorCode["ColonExpected"] = 5)] = "ColonExpected";
  ParseErrorCode[(ParseErrorCode["CommaExpected"] = 6)] = "CommaExpected";
  ParseErrorCode[(ParseErrorCode["CloseBraceExpected"] = 7)] = "CloseBraceExpected";
  ParseErrorCode[(ParseErrorCode["CloseBracketExpected"] = 8)] = "CloseBracketExpected";
  ParseErrorCode[(ParseErrorCode["EndOfFileExpected"] = 9)] = "EndOfFileExpected";
  ParseErrorCode[(ParseErrorCode["InvalidCommentToken"] = 10)] = "InvalidCommentToken";
  ParseErrorCode[(ParseErrorCode["UnexpectedEndOfComment"] = 11)] = "UnexpectedEndOfComment";
  ParseErrorCode[(ParseErrorCode["UnexpectedEndOfString"] = 12)] = "UnexpectedEndOfString";
  ParseErrorCode[(ParseErrorCode["UnexpectedEndOfNumber"] = 13)] = "UnexpectedEndOfNumber";
  ParseErrorCode[(ParseErrorCode["InvalidUnicode"] = 14)] = "InvalidUnicode";
  ParseErrorCode[(ParseErrorCode["InvalidEscapeCharacter"] = 15)] = "InvalidEscapeCharacter";
  ParseErrorCode[(ParseErrorCode["InvalidCharacter"] = 16)] = "InvalidCharacter";
})(ParseErrorCode || (ParseErrorCode = {}));
//#endregion
//#region src/utils/json.ts
function readJsonFile(file, allowComments) {
  const content = fs.readFileSync(file, "utf-8");
  return (allowComments ? parse : JSON.parse)(content);
}
function writeJsonFile(file, data) {
  let newline = "\n";
  let indent = "  ";
  if (fs.existsSync(file)) {
    const content = fs.readFileSync(file, "utf-8");
    indent = detectIndent(content).indent;
    newline = detectNewline(content) ?? "";
  }
  fs.writeFileSync(file, JSON.stringify(data, null, indent) + newline, "utf-8");
}
function editJsonFile(file, callback) {
  const newJson = callback(readJsonFile(file));
  if (newJson) writeJsonFile(file, newJson);
}
function isJsonFile(file) {
  try {
    readJsonFile(file);
    return true;
  } catch {
    return false;
  }
}
//#endregion
//#region src/utils/package.ts
function getScopeFromPackageName(packageName) {
  if (packageName.startsWith("@")) return packageName.split("/")[0];
  return "";
}
function detectPackageMetadata(projectPath, packageName) {
  try {
    const pkgFilePath = createRequire(path.join(projectPath, "noop.js")).resolve(
      `${packageName}/package.json`,
    );
    const pkg = JSON.parse(fs.readFileSync(pkgFilePath, "utf8"));
    return {
      name: pkg.name,
      version: pkg.version,
      path: path.dirname(pkgFilePath),
    };
  } catch {
    return;
  }
}
/**
 * Read the nearest package.json file from the current directory up to the root directory.
 * @param currentDir - The current directory to start searching from.
 * @returns The package.json content as a JSON object, or null if no package.json is found.
 */
function readNearestPackageJson(currentDir) {
  do {
    const packageJsonPath = path.join(currentDir, "package.json");
    if (fs.existsSync(packageJsonPath)) return readJsonFile(packageJsonPath);
    currentDir = path.dirname(currentDir);
  } while (currentDir !== path.dirname(currentDir));
  return null;
}
function hasVitePlusDependency(pkg) {
  return Boolean(pkg?.dependencies?.["vite-plus"] || pkg?.devDependencies?.["vite-plus"]);
}
//#endregion
export {
  editJsonFile as a,
  writeJsonFile as c,
  VITE_PLUS_OVERRIDE_PACKAGES as d,
  VITE_PLUS_VERSION as f,
  readNearestPackageJson as i,
  BASEURL_TSCONFIG_WARNING as l,
  getScopeFromPackageName as n,
  isJsonFile as o,
  hasVitePlusDependency as r,
  readJsonFile as s,
  detectPackageMetadata as t,
  VITE_PLUS_NAME as u,
};
